File: Astromaximum/Classes/utils/AmaxLocationDataFile.m
//
//  AmaxLocationDataFile.m
//  Astromaximum
//
//  Created by admin on 14.11.12.
//  Copyright (c) 2012 S&W Axis. All rights reserved.
//

#import "AmaxLocationDataFile.h"
#import "AmaxDataInputStream.h"
#import "AmaxTimezoneTransition.h"

@implementation AmaxLocationDataFile

@synthesize mStartYear = _mStartYear;
@synthesize mStartMonth = _mStartMonth;
@synthesize mStartDay = _mStartDay;
@synthesize mDayCount = _mDayCount;
@synthesize mCityId = _mCityId;
@synthesize mCity = _mCity;
@synthesize mState = _mState;
@synthesize mCountry = _mCountry;
@synthesize mTimezone = _mTimezone;
@synthesize mData = _mData;

- (id)initWithBytes:(const void *)bytes length:(NSUInteger)length headerOnly:(BOOL)headerOnly
{
    AmaxDataInputStream *dis = [[AmaxDataInputStream alloc]initWithBytes:bytes length:length];
    [dis skipBytes:4]; // signature
    char version = [dis readUnsignedByte];
    _mStartYear = [dis readShort];
    _mStartMonth = [dis readUnsignedByte] - 1;
    _mStartDay = [dis readUnsignedByte];
    if (version == 3) {
        _mMonthCount = [dis readUnsignedByte];
        _mDayCount = 365;
    }
    else if (version == 2) {
        _mDayCount = [dis readShort];
        _mMonthCount = 12;
    }
    else {
        NSLog(@"Unknown version %d", version);
    }
    _mCityId = [dis readInt]; // city id
    coords[0] = [dis readShort]; // latitude
    coords[1] = [dis readShort]; // longitude
    coords[2] = [dis readShort]; // altitude
    _mCity = [dis readUTF]; // city
    _mState = [dis readUTF]; // state
    _mCountry = [dis readUTF]; // country
    _mTimezone = [dis readUTF]; // timezone
    if (headerOnly == NO) {
        customData = [dis readUTF]; // custom data
        int transitionCount = [dis readUnsignedByte];
        transitions = [NSMutableArray array];
        for (int i = 0; i < transitionCount; ++i) {
            NSDate *time = [NSDate dateWithTimeIntervalSince1970:[dis readInt]]; // start_date
            NSTimeInterval offset = [dis readShort] * 60; // gmt_ofs_min
            NSString* name = [dis readUTF]; // name
            AmaxTimezoneTransition *transition = [[AmaxTimezoneTransition alloc]initWithTime:time offset:offset name:name];
            [transitions addObject:transition];
        }
        Size bufferLength = [dis availableBytes];
        char *buffer = malloc(bufferLength);
        [dis readToBuffer:buffer length:bufferLength];
        _mData = [[AmaxDataInputStream alloc]initWithBytes:buffer length:bufferLength];
    }
    return self;
}

@end

File: Astromaximum/Classes/utils/AmaxLocationDataFile.swift
//
//  AmaxLocationDataFile.m
//  Astromaximum
//
//  Created by admin on 14.11.12.
//  Copyright (c) 2012 S&W Axis. All rights reserved.
//

import Foundation

class AmaxLocationDataFile {

    private var customData: String?
    private var transitions: [AmaxTimezoneTransition]?

    var mStartYear: Int
    var mStartMonth: Int
    var mStartDay: Int
    var mMonthCount: Int = 0
    var location = AmaxLocation()

    var mData: AmaxDataInputStream?

    init(data: Data, headerOnly: Bool) {
        let dis = AmaxDataInputStream(data:data)
        dis.skipBytes(4) // signature
        let version = dis.readUnsignedByte()
        mStartYear = Int(dis.readShort())
        mStartMonth = Int(dis.readUnsignedByte() - 1)
        mStartDay = Int(dis.readUnsignedByte())
        if version == 3 {
            mMonthCount = Int(dis.readUnsignedByte())
        }
        else if version == 2 {
            mMonthCount = 12
        }
        else {
            NSLog("Unknown version %d", version)
        }

        location.mCityId = String(format: "%X", dis.readInt()) // city id
        location.mCoords[0] = Int(dis.readShort()) // latitude
        location.mCoords[1] = Int(dis.readShort()) // longitude
        location.mCoords[2] = Int(dis.readShort()) // altitude
        location.mCity = dis.readUTF()! // city
        location.mState = dis.readUTF()! // state
        location.mCountry = dis.readUTF()! // country
        location.mTimezone = dis.readUTF()! // timezone

        if headerOnly == false {
            customData = dis.readUTF() // custom data
            let transitionCount = dis.readUnsignedByte()
            transitions = [AmaxTimezoneTransition]()
            for _ in 0 ..< transitionCount {
                let time = Date(timeIntervalSince1970: TimeInterval(dis.readInt())) // start_date
                let offset = TimeInterval(dis.readShort() * 60) // gmt_ofs_min
                let name = dis.readUTF()! // name
                let transition = AmaxTimezoneTransition(time:time, offset:offset, name:name)
                transitions!.append(transition)
             }
            let bufferLength = dis.availableBytes()
            let data = dis.read(length: bufferLength)
            mData = AmaxDataInputStream(data:data)
        }
    }
}
